Whats new in .NET 11 preview 7 (#2274)
dotnet/docs-desktop: dotnet-desktop-guide/winforms/whats-new/net110.md

<xref:System.Windows.Forms.ComboBox>, <xref:System.Windows.Forms.ListBox>, <xref:System.Windows.Forms.ListView>, <xref:System.Windows.Forms.RichTextBox>, and <xref:System.Windows.Forms.TreeView> now implement `ISupportSuspendPainting`. They route their established suspension patterns (`BeginUpdate`/`EndUpdate`) through the new interface, so bulk edits avoid flicker without managing multiple paired calls. For all other controls, the suspension behavior is new.

```csharp
using System.Windows.Forms;

ListBox list = new();

using (list.SuspendPainting(LayoutSuspendTraversal.Target))
{
  for (int i = 0; i < 10_000; i++)
  {
    list.Items.Add($"Item {i}");
  }
}
```

## Toggle-switch appearance for CheckBox and RadioButton

> `System.Windows.Forms.Appearance.ToggleSwitch` is a new value for `System.Windows.Forms.CheckBox.Appearance` that renders the control as a switch under `VisualStylesMode.Net11`. Existing `Checked`, `CheckedChanged`, and data-binding behavior stays unchanged, so you don't need code updates when switching a `CheckBox` to the new appearance. Modernize checkbox controls to align with contemporary UI conventions without breaking changes.

Reviewer: ```suggestion
`System.Windows.Forms.Appearance.ToggleSwitch` is a new value for <xref:System.Windows.Forms.CheckBox.Appearance?displayProperty=nameWithType> that renders the control as a switch under `VisualStylesMode.Net11`. Existing `Checked`, `CheckedChanged`, and data-binding behavior stays unchanged, so you don't need code updates when switching a `CheckBox` to the new appearance. You can modernize checkbox controls to align with contemporary UI conventions without encountering breaking changes.
```

```csharp
using System.Windows.Forms;

CheckBox toggle = new()
{
  Text = "Enable notifications",
  Appearance = Appearance.ToggleSwitch,
  Checked = true,
};
```

> [!IMPORTANT]

> > Rendering behavior for Visual Styles and the associated new APIs is not yet final and will continue to change until .NET 11 reaches GA. Based on exploratory testing and customer feedback, controls using non-classic Visual Style settings may occupy more or less space in upcoming .NET 11 releases (RC, GA) than they do in Preview 7. The behavior of the new APIs may likewise change in detail or in specific scenarios.
> >
> > For this reason we recommend avoiding pixel-perfect design when adopting these features. Prefer layout-based approaches using `TableLayoutPanel` and `FlowLayoutPanel`, which adapt automatically to changes in control metrics.

Reviewer: ```suggestion
> Rendering behavior for Visual Styles and the associated new APIs is not yet final and will continue to change until .NET 11 reaches GA. Based on exploratory testing and customer feedback, controls that use non-classic Visual Style settings might occupy more or less space in upcoming .NET 11 releases (RC, GA) than they do in Preview 7. The behavior of the new APIs might likewise change in detail or in specific scenarios.
>
> For this reason, we recommend avoiding pixel-perfect design when adopting these features. Prefer layout-based approaches using `TableLayoutPanel` and `FlowLayoutPanel`, which adapt automatically to changes in control metrics.
```

## Dark mode

- <xref:System.Windows.Forms.ProgressBar> now has default foreground and background colors for dark mode rendering, improving visual consistency in Fluent-themed applications.

## Bug fixes

- <xref:System.Windows.Forms.Control>: The <xref:System.Windows.Forms.SendKeys> sequential immediate modifier parsing (such as `^a^c`, `+a+c`, `%f%t`) regression is fixed. The `KeyboardToolTipStateMachine` no longer throws `KeyNotFoundException` when focus changes during a tooltip popup event.

- <xref:System.Windows.Forms.ListBox>: `IsSynchronized` values are now correct across `ListBox` collections.

- <xref:System.Windows.Forms.NumericUpDown>: Button rendering regression when visual styles are disabled is fixed.

- <xref:System.Windows.Forms.SplitContainer>: Transient GDI+ `ExternalException` during display-session transition when calling `RepaintSplitterRect` is now handled.

- <xref:System.Windows.Forms.ToolStrip>: The scroll-down button no longer throws when scrolling. Indeterminate and checked `ToolStripMenuItem` icons are now visible in dark mode.

> - <xref:System.Windows.Forms.PropertyGrid>: Property Page button icon no longer changes when `LargeButtons` is toggled on high DPI. `ResetHelpForeColor` now correctly resets the foreground color instead of the background color.

Reviewer: ```suggestion
- <xref:System.Windows.Forms.PropertyGrid>: The Property Page button icon no longer changes when `LargeButtons` is toggled on high DPI. `ResetHelpForeColor` now correctly resets the foreground color instead of the background color.
```
